R/sample.R

```diff
@@ -97,3 +103,100 @@ mcstate_sampler <- function(name, initialise, step, finalise) {
   class(ret) <- "mcstate_sampler"
   ret
 }
+
+
+initial_parameters <- function(initial, model, rng, call = NULL) {
+  n_pars <- length(model$parameters)
+  n_chains <- length(rng)
+  if (is.null(initial)) {
+    ## Really this would just be from the prior; we can't directly
+    ## sample from the posterior!
+    initial <- lapply(rng, function(r) model$direct_sample(r))
+  }
+  if (is.list(initial)) {
+    if (length(initial) != n_chains) {
+      cli::cli_abort(
+        c(paste("Unexpected length for list 'initial'",
+                "(given {length(initial)}, expected {n_chains})"),
+          i = paste("You are running {n_chains}, so I expected a list with",
+                    "that many entries, each of which has {n_pars}",
+                    "element{?s}")),
+        arg = "initial", call = call)
+    }
+    err <- lengths(initial) != n_pars
+    if (any(err)) {
+      ## We might want to highlight the entries that are bad, and also
+      ## indicate if it was direct_sample() that produced the invalid
+      ## entry.
+      cli::cli_abort(
+        c(paste("Unexpected initial parameter length; expected {n_pars}"),
+          i = "Incorrect length for element{?s} {as.character(which(err))}",
+          i = paste("Your model has {n_pars} parameter{?s}, so each",
+                    "element of your list must have this many elements")),
+        arg = "initial", call = call)
+    }
+    initial <- matrix(unlist(unname(initial)), n_chains, n_pars, byrow = TRUE)
+  } else if (is.matrix(initial)) {
+    hint_matrix <- paste(
+      "Expected a matrix with {n_chains} row{?s} (1 per chain) and",
+      "{n_pars} column{?s} (1 per parameter)")
+    if (nrow(initial) != n_chains) {
+      cli::cli_abort(
+        c(paste("Unexpected number of rows in 'initial'",
+                "(given {nrow(initial)}, expected {n_chains})"),
+          i = hint_matrix),
+        arg = "initial", call = call)
+    }
+    if (ncol(initial) != n_pars) {
+      cli::cli_abort(
+        c(paste("Unexpected number of columns in 'initial'",
+                "(given {ncol(initial)}, expected {n_pars})"),
+          i = hint_matrix),
+        arg = "initial", call = call)
+    }
+  } else {
+    if (length(initial) != n_pars) {
+      cli::cli_abort(
+        c(paste("Unexpected length for vector 'initial'",
+                "(given {length(initial)}, expected {n_pars})"),
+          i = "I expected a vector with {n_pars} element{?s}, 1 per parameter"),
+        arg = "initial", call = call)
+    }
+    initial <- matrix(initial, n_chains, n_pars, byrow = TRUE)
+  }
+  initial
+}
+
+
+combine_chains <- function(res) {
+  stopifnot(is.null(res$chain))
+  chain <- rep(seq_along(res), lengths(lapply(res, "[[", "density")))
+
+  pars <- rbind_list(lapply(res, "[[", "pars"))
+  density <- unlist(lapply(res, "[[", "density"))
```

Comment:
just for clarification, why is pars a matrix and density a vector here? is it because we expect multiple parameters in the pars so making it a matrix is better but density will be `1 x n` where `n` is number of samples?

Reply:
density is a mapping from pars -> real, with pars on R^n.  So in the history we have n_steps densities but pars is a n_steps x length(pars) matrix.